File: fastapi-python-backend/src/errors/error_map.py
from json import JSONDecodeError

from fastapi.exceptions import RequestValidationError
from pydantic import ValidationError as PydanticValidationError

from .app_error import AppError
from .http_errors import InternalServerError, ValidationError


def map_to_app_error(error: Exception | BaseException | object) -> AppError:
    if isinstance(error, AppError):
        return error

    if isinstance(error, (PydanticValidationError, RequestValidationError)):
        return ValidationError(
            detail='Request validation failed.',
            errors=list(error.errors()),
            cause=error if isinstance(error, Exception) else None,
        )

    if isinstance(error, JSONDecodeError):
        return ValidationError(
            detail='Malformed JSON body.',
            cause=error,
        )

    if isinstance(error, Exception):
        return InternalServerError(
            detail='Unexpected server error.',
            cause=error,
            meta={
                'originalErrorName': error.__class__.__name__,
            },
        )

    return InternalServerError(
        detail='Unexpected server error.',
        meta={
            'unknownErrorType': type(error).__name__,
        },
    )

File: hono-ts-bun-backend/src/errors/error-map.ts
import { ZodError } from "zod";
import { AppError } from "./app-error";
import { InternalServerError, RateLimitExceededError, ValidationError } from "./http-errors";

type GenericError = Error & {
  status?: number;
  statusCode?: number;
  code?: string;
};

export function mapToAppError(error: unknown): AppError {
  if (error instanceof AppError) {
    return error;
  }

  if (error instanceof ZodError) {
    return new ValidationError({
      detail: "Request validation failed.",
      errors: error.issues,
      cause: error,
    });
  }

  if (error instanceof SyntaxError) {
    return new ValidationError({
      detail: "Malformed JSON body.",
      cause: error,
    });
  }

  if (error instanceof Error) {
    const candidate = error as GenericError;
    if (
      candidate.status === 429 ||
      candidate.statusCode === 429 ||
      candidate.code === "RATE_LIMIT"
    ) {
      return new RateLimitExceededError({
        detail: "Too many requests, please try again later.",
        cause: error,
      });
    }

    return new InternalServerError({
      detail: "Unexpected server error.",
      cause: error,
      meta: {
        originalErrorName: error.name,
      },
    });
  }

  return new InternalServerError({
    detail: "Unexpected server error.",
    meta: {
      unknownErrorType: typeof error,
    },
  });
}
